(0.13.54) Certain scenario names seem to crash scenario service

<!--

This issue tracker is only for technical issues related to the DAML SDK.

If you suspect that your issue has a security impact, do not use this issue tracker. In that case, please contact security@digitalasset.com and follow our responsible disclosure policy: https://digitalasset.com/security/ .

For general questions about DAML and/or support request you should use Stack Overflow: https://stackoverflow.com/questions/tagged/daml .

-->

I've found that scenario names containing an apostrophe in the middle will sometimes consistently compile and cause scenario service backend errors like the following:

```
2020-04-17 06:45:00.59 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR: Apr 17, 2020 4:45:00 PM io.grpc.internal.SerializingExecutor run

2020-04-17 06:45:00.59 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR: SEVERE: Exception while executing runnable io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed@47849443

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR: com.digitalasset.daml.lf.archive.Reader$ParseError: string "se't_test" does not match regex "[A-Za-z\$_][A-Za-z0-9\$_]*"

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at com.digitalasset.daml.lf.scenario.Context$.$anonfun$assert$1(Context.scala:42)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at scala.util.Either.fold(Either.scala:189)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at com.digitalasset.daml.lf.scenario.Context$.com$digitalasset$daml$lf$scenario$Context$$assert(Context.scala:42)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at com.digitalasset.daml.lf.scenario.Context.interpretScenario(Context.scala:172)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at com.digitalasset.daml.lf.scenario.ScenarioService.$anonfun$runScenario$1(ScenarioServiceMain.scala:86)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at scala.Option.flatMap(Option.scala:171)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at com.digitalasset.daml.lf.scenario.ScenarioService.runScenario(ScenarioServiceMain.scala:74)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at com.digitalasset.daml.lf.scenario.api.v1.ScenarioServiceGrpc$MethodHandlers.invoke(ScenarioServiceGrpc.java:643)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:808)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

2020-04-17 06:45:00.60 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR:        at java.lang.Thread.run(Thread.java:748)

2020-04-17 06:45:00.61 [ERROR]  [test ScenarioService]  [] 
SCENARIO SERVICE STDERR: 
File:     src/Test/My/Module.daml
Hidden:   no
Range:    62:0-62:9
Source:   Scenario
Severity: DsError
Message: 
  "Scenario service backend error: BErrorClient (ClientIOError (GRPCIOBadStatusCode StatusUnknown
  (StatusDetails {unStatusDetails = \"\"})))"
```

In this case, `se't_test` was the name of the problematic scenario.

--- a/compiler/damlc/tests/daml-test-files/MangledScenario'.daml
+++ b/compiler/damlc/tests/daml-test-files/MangledScenario'.daml
@@ -1,0 +1,13 @@
+module MangledScenario' where
+
+template T'
+  with
+    p : Party
+  where
+    signatory p
+
+mangled' = scenario do
+  alice <- getParty "Alice"
+  t' <- submit alice (create T' with p = alice)
+  submit alice (exercise t' Archive)
+  pure ()

--- a/compiler/damlc/tests/src/DA/Test/ShakeIdeClient.hs
+++ b/compiler/damlc/tests/src/DA/Test/ShakeIdeClient.hs
@@ -247,6 +247,25 @@ basicTests mbScenarioService = Tasty.testGroup "Basic tests"
             expectVirtualResource va "Return value: &quot;foo&quot;"
             expectVirtualResource vb "Return value: &quot;bar&quot;"
 
+    , testCase' "Scenario with mangled names" $ do
+            a <- makeFile "foo/MangledScenario'.daml" $ T.unlines
+                [ "module MangledScenario' where"
+                , "template T' with"
+                , "    p : Party"
+                , "  where"
+                , "    signatory p"
+                , "mangled' = scenario do"
+                , "  alice <- getParty \"Alice\""
+                , "  t' <- submit alice (create (T' alice))"
+                , "  submit alice (exercise t' Archive)"
+                ]
+            setFilesOfInterest [a]
+            expectNoErrors
+            let va = VRScenario a "mangled'"
+            setOpenVirtualResources [va]
+            expectVirtualResource va "title=\"MangledScenario':T'\""
+
+
     ,   testCaseFails' "Modules must match their filename DEL-7175" $ do
             a <- makeFile "Foo/Test.daml" "daml 1.2 module Test where"
             setFilesOfInterest [a]

--- a/compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs
+++ b/compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs
@@ -1,6 +1,6 @@
 -- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 -- SPDX-License-Identifier: Apache-2.0
-
+{-# LANGUAGE PatternSynonyms #-}
 
 -- | Pretty-printing of scenario results
 module DA.Daml.LF.PrettyScenario
@@ -18,6 +18,7 @@ import           Control.Monad
 import           Control.Monad.Reader
 import           Control.Monad.Trans.Except
 import qualified DA.Daml.LF.Ast             as LF
+import DA.Daml.LF.Mangling
 import Control.Applicative
 import Text.Read hiding (parens)
 import           DA.Pretty as Pretty
@@ -45,6 +46,23 @@ data Error = ErrorMissingNode NodeId
 type M = ExceptT Error (Reader (MS.Map NodeId Node, LF.World))
 
 type ModuleRef = LF.Qualified ()
+
+unmangleQualifiedName :: T.Text -> (LF.ModuleName, T.Text)
+unmangleQualifiedName t = case T.splitOn ":" t of
+    [modName, defName] -> (unmangleModuleName modName, unmangle defName)
+    _ -> error "Bad definition"
+
+unmangleModuleName :: T.Text -> LF.ModuleName
+unmangleModuleName t = LF.ModuleName (map unmangle $ T.splitOn "." t)
+
+unmangle :: T.Text -> T.Text
+unmangle s = case unmangleIdentifier s of
+    Left err -> error err
+    Right (UnmangledIdentifier s) -> s
+
+{-# COMPLETE UnmangledQualifiedName #-}
+pattern UnmangledQualifiedName :: LF.ModuleName -> T.Text -> TL.Text
+pattern UnmangledQualifiedName mod def <- (unmangleQualifiedName . TL.toStrict -> (mod, def))
 
 runM :: V.Vector Node -> LF.World -> M (Doc SyntaxClass) -> Doc SyntaxClass
 runM nodes world =
@@ -82,16 +100,6 @@ lookupModule world mbPkgId modName = do
          LF.PRImport $ LF.PackageId $ TL.toStrict pkgId
        _ -> LF.PRSelf
   eitherToMaybe (LF.lookupModule (LF.Qualified pkgRef modName ()) world)
-
-lookupModuleFromQualifiedName ::
-     LF.World -> Maybe PackageIdentifier -> T.Text
-  -> Maybe (LF.ModuleName, T.Text, LF.Module)
-lookupModuleFromQualifiedName world mbPkgId qualName = do
-  let (modName, defName) = case T.splitOn ":" qualName of
-        [modNm, defNm] -> (LF.ModuleName (T.splitOn "." modNm), defNm)
-        _ -> error "Bad definition"
-  modu <- lookupModule world mbPkgId modName
-  return (modName, defName, modu)
 
 parseNodeId :: NodeId -> [Integer]
 parseNodeId =
@@ -428,12 +436,12 @@ prettyMayLocation :: LF.World -> Maybe Location -> Doc SyntaxClass
 prettyMayLocation world = maybe (text "unknown source") (prettyLocation world)
 
 prettyLocation :: LF.World -> Location -> Doc SyntaxClass
-prettyLocation world (Location mbPkgId modName sline scol eline _ecol _definition) =
+prettyLocation world (Location mbPkgId (unmangleModuleName . TL.toStrict -> modName) sline scol eline _ecol _definition) =
       maybe id (\path -> linkSC (url path) title)
-        (lookupModule world mbPkgId (LF.ModuleName (T.splitOn "." (TL.toStrict modName))) >>= LF.moduleSource)
+        (lookupModule world mbPkgId modName >>= LF.moduleSource)
     $ text title
   where
-    modName' = TL.toStrict modName
+    modName' = LF.moduleNameString modName
     encodeURI = Network.URI.Encode.encodeText
     title = modName' <> lineNum
     url fp = "command:daml.revealLocation?"
@@ -710,15 +718,15 @@ prettyPackageIdentifier (PackageIdentifier psum) = case psum of
   (Just (PackageIdentifierSumPackageId pid)) -> char '@' <> ltext pid
 
 prettyDefName :: LF.World -> Identifier -> Doc SyntaxClass
-prettyDefName world (Identifier mbPkgId (TL.toStrict -> qualName))
-  | Just (_modName, defName, mod0) <- lookupModuleFromQualifiedName world mbPkgId qualName
+prettyDefName world (Identifier mbPkgId (UnmangledQualifiedName modName defName))
+  | Just mod0 <- lookupModule world mbPkgId modName
   , Just fp <- LF.moduleSource mod0
   , Just (LF.SourceLoc _mref sline _scol eline _ecol) <- lookupDefLocation mod0 defName =
       linkSC (revealLocationUri fp sline eline) name ppName
   | otherwise =
       ppName
   where
-    name = qualName
+    name = LF.moduleNameString modName <> ":" <> defName
     ppName = text name <> ppPkgId
     ppPkgId = maybe mempty prettyPackageIdentifier mbPkgId
 
@@ -726,8 +734,8 @@ prettyChoiceId
   :: LF.World -> Maybe Identifier -> TL.Text
   -> Doc SyntaxClass
 prettyChoiceId _ Nothing choiceId = ltext choiceId
-prettyChoiceId world (Just (Identifier mbPkgId (TL.toStrict -> qualName))) (TL.toStrict -> choiceId)
-  | Just (_modName, defName, mod0) <- lookupModuleFromQualifiedName world mbPkgId qualName
+prettyChoiceId world (Just (Identifier mbPkgId (UnmangledQualifiedName modName defName))) (TL.toStrict -> choiceId)
+  | Just mod0 <- lookupModule world mbPkgId modName
   , Just fp <- LF.moduleSource mod0
   , Just tpl <- NM.lookup (LF.TypeConName [defName]) (LF.moduleTemplates mod0)
   , Just chc <- NM.lookup (LF.ChoiceName choiceId) (LF.tplChoices tpl)

--- a/compiler/scenario-service/client/src/DA/Daml/LF/ScenarioServiceClient/LowLevel.hs
+++ b/compiler/scenario-service/client/src/DA/Daml/LF/ScenarioServiceClient/LowLevel.hs
@@ -27,6 +27,7 @@ module DA.Daml.LF.ScenarioServiceClient.LowLevel
   ) where
 
 import Conduit (runConduit, (.|), MonadUnliftIO(..))
+import Data.Either
 import Data.Maybe
 import Data.IORef
 import GHC.Generics
@@ -37,6 +38,7 @@ import Control.DeepSeq
 import Control.Exception
 import Control.Monad
 import Control.Monad.IO.Class
+import DA.Daml.LF.Mangling
 import qualified DA.Daml.LF.Proto3.EncodeV1 as EncodeV1
 import qualified Data.ByteString as BS
 import qualified Data.ByteString.Lazy as BSL
@@ -329,10 +331,16 @@ runScenario Handle{..} (ContextId ctxId) name = do
       let ssPkgId = SS.PackageIdentifier $ Just $ case pkgId of
             LF.PRSelf     -> SS.PackageIdentifierSumSelf SS.Empty
             LF.PRImport x -> SS.PackageIdentifierSumPackageId (TL.fromStrict $ LF.unPackageId x)
+          mangledDefn =
+              fromRight (error "Failed to mangle scenario name") $
+              mangleIdentifier (LF.unExprValName defn)
+          mangledModName = T.intercalate "." $
+              map (fromRight (error "Failed to mangle scenario module name") . mangleIdentifier) (LF.unModuleName modName)
+
       in
         SS.Identifier
           (Just ssPkgId)
-          (TL.fromStrict $ T.intercalate "." (LF.unModuleName modName) <> ":" <> LF.unExprValName defn)
+          (TL.fromStrict $ mangledModName <> ":" <> mangledDefn)
 
 performRequest
   :: (ClientRequest 'Normal payload response -> IO (ClientResult 'Normal response))

--- a/compiler/scenario-service/client/src/DA/Daml/LF/ScenarioServiceClient/LowLevel.hs
+++ b/compiler/scenario-service/client/src/DA/Daml/LF/ScenarioServiceClient/LowLevel.hs
@@ -307,11 +307,16 @@ updateCtx Handle{..} (ContextId ctxId) ContextUpdate{..} = do
       SS.UpdateContextRequest_UpdatePackages
         (V.fromList (map snd updLoadPackages))
         (V.fromList (map (TL.fromStrict . LF.unPackageId) updUnloadPackages))
-    encodeName = TL.fromStrict . T.intercalate "." . LF.unModuleName
+    encodeName = TL.fromStrict . mangleModuleName
     convModule :: (LF.ModuleName, BS.ByteString) -> SS.ScenarioModule
     convModule (_, bytes) =
         case updDamlLfVersion of
             LF.V1 minor -> SS.ScenarioModule bytes (TL.pack $ LF.renderMinorVersion minor)
+
+mangleModuleName :: LF.ModuleName -> T.Text
+mangleModuleName (LF.ModuleName modName) =
+    T.intercalate "." $
+    map (fromRight (error "Failed to mangle scenario module name") . mangleIdentifier) modName
 
 runScenario :: Handle -> ContextId -> LF.ValueRef -> IO (Either Error SS.ScenarioResult)
 runScenario Handle{..} (ContextId ctxId) name = do
@@ -334,9 +339,7 @@ runScenario Handle{..} (ContextId ctxId) name = do
           mangledDefn =
               fromRight (error "Failed to mangle scenario name") $
               mangleIdentifier (LF.unExprValName defn)
-          mangledModName = T.intercalate "." $
-              map (fromRight (error "Failed to mangle scenario module name") . mangleIdentifier) (LF.unModuleName modName)
-
+          mangledModName = mangleModuleName modName
       in
         SS.Identifier
           (Just ssPkgId)

